src/starks/traits.rs

```diff
@@ -95,33 +95,25 @@ pub trait AIR: Clone {
 
     fn pub_inputs(&self) -> &Self::PublicInputs;
 
-    fn transition_exemptions_verifier(&self) -> Vec<Polynomial<FieldElement<Self::Field>>> {
-        let trace_length = self.trace_length();
-        let roots_of_unity_order = trace_length.trailing_zeros();
-        let roots_of_unity = get_powers_of_primitive_root_coset(
-            roots_of_unity_order as u64,
-            self.trace_length(),
-            &FieldElement::<Self::Field>::one(),
-        )
-        .unwrap();
-        let root_of_unity_len = roots_of_unity.len();
-
+    fn transition_exemptions_verifier(
+        &self,
+        root: &FieldElement<Self::Field>,
+    ) -> Vec<Polynomial<FieldElement<Self::Field>>> {
+        //let trace_length = self.trace_length();
         let x = Polynomial::new_monomial(FieldElement::one(), 1);
 
-        self.context()
+        let max = self
+            .context()
             .transition_exemptions
             .iter()
-            .take(self.context().num_transition_constraints)
-            .map(|cant_take| {
-                roots_of_unity
-                    .iter()
-                    .take(root_of_unity_len)
-                    .rev()
-                    .take(*cant_take)
-                    .fold(
-                        Polynomial::new_monomial(FieldElement::one(), 0),
-                        |acc, root| acc * (&x - root),
-                    )
+            .max()
+            .expect("has maximum");
+        (1..=*max)
+            .map(|index| {
+                (1..=index).fold(
+                    Polynomial::new_monomial(FieldElement::one(), 0),
+                    |acc, k| acc * (&x - root.pow(k)),
+                )
             })
```

Comment:
I think we could replace the `root.pow(k)` by keeping the current product as part of the `fold` accumulator. It would probably look something like this:
```suggestion
            .map(|index| {
                (1..=index).fold(
                    (Polynomial::new_monomial(FieldElement::one(), 0), root.clone()),
                    |acc, k| (acc.0 * (&x - acc.1), acc.1 * root),
                )
            })
```

Reply:
Actually, this introduces a problem when recovering the results. An even better way would be to use scan or something like that. I just reverted to the previous version